Feature/verona 3.0 complements

jojohoch opened the pull request:
Player implementiert fehlende [Verona 3.0 Spezifikationen](https://github.com/verona-interfaces/player) :
- **vopNavigationDeniedNotification** Nachricht von Host zu Player
Der Player subscribed sich beim `EventService`, um informiert zu werden, wenn das Event ausgelöst wird. Daraufhin wird überprüft, ob das Formular des Players valide ist und Fehler dem User ggfs. angezeigt.
- **vopStateChangedNotification** (type: presentationProgress, detail: complete) Nachricht von Player zu Host
Der Player subscribed sich beim `EventService`, um Scrollinformationen des Users zu erhalten. Diese werden genutzt, um zu berechnen, ob das gesamte Formular vom User gesehen wurde. Wenn ja, wird die Information an den Host gesendet.

Tests:
- `EventService`
- `ParserService` Die Tests des ehemaligen `DataService` wurden an die überarbeiteten `UIBlocks` und die neuen Methoden des `ParserService` angepasst

Commit:
Rework tests for parser service

Make use of
* the renamed parser service (formerly data service)
* its new methods
* the reworked structure of UIBlocks

Commit:
Add parserOptions for eslint configuration in package.json

Commit:
Remove unnecessary log to console

Commit:
Implement `vopNavigationDeniedNotification` handling

To receive and handle the event the player subscribes to an
`EventService` which is listening to the
`vopNavigationDeniedNotification` notification sent by the host.

Commit:
Implement info that presentation is complete

When a user has seen the complete content of the form the player posts
a message to the host.

rhenck commented:
"Add parserOptions for eslint configuration in package.json"

Der Commit hat ja eigentlich gar nichts mit dem Pull Request und Verona 3 zu tun. Das hätte ich extra gemacht. Und sowas kannst von mir aus auch direkt, ohne weitere Absprachen,  auf den master pushen. Da gibts ja keine 2 Meinungen.

Edit: Das gleiche gilt ja im Prinzip auch für die Tests und für "Remove unnecessary log to console".

rhenck commented on src/app/event.service.ts:
```diff
@@ -0,0 +1,26 @@
+      .subscribe(() => this.scrollY.next(window.scrollY));
+  }
+
+  get navigationDenied$(): Observable<string[]> {
```
Wir haben uns ja beim Aspect darauf geeinigt, das mit dem $ sein zu lassen. Wollen wir das hier auch weglassen?

jojohoch replied:
umbennant!

rhenck commented on src/app/player.component.ts:
```diff
@@ -20,25 +23,49 @@ import { InputElement } from './classes/UIElement';
+      .subscribe((): void => this.form.markAllAsTouched());
+    this.eventService.scrollY$
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((y: number): void => this.calculatePresentationComplete(y));
```
y finde ich hier grenzwertig. Kann man machen, aber generell besser mehr als 1 Buchstaben für eine Variable. (Ausnahme 'i').

jojohoch replied:
umbenannt

rhenck commented on src/app/player.component.ts:
```diff
@@ -20,25 +23,49 @@ import { InputElement } from './classes/UIElement';
+  }
+
+  private calculatePresentationComplete(y: number): void {
+    const contentPos = window.innerHeight + y;
```
Hier auch: "Pos" ist mir nicht sofort klar.

jojohoch replied:
umbenannt

rhenck commented on src/app/player.component.ts:
```diff
@@ -50,19 +77,22 @@ export class PlayerComponent {
       }
       this.rootBlock = this.parserService.parseUnitDefinition(startData.unitDefinition.split(/\r?\n/g));
       this.rootBlock.check(storedResponses);
+      // check if presentationProgress could be sent (e.g. small pages which hasn't to be scrolled)
```
Kleiner Rechtschreibfehler. :)

Entweder "small page which hasn't to be scrolled" oder "small pages which haven't to be scrolled"

jojohoch replied:
fixed

rhenck reviewed (approved):
Paar Kleinigkeiten, aber nicht schlimmes. Kannst du noch verbessern, aber von mir aus auch so mergen.

jojohoch committed:
Fix typo

`unitStateType` should be `unitStateDataType` as described in Verona
Interfaces Specification

jojohoch committed:
Rename observables

Remove '$' at the end of the property name

jojohoch committed:
Rename variable

Rename 'y' to 'scrolledY'

jojohoch committed:
Rename variable

Rename `contentPos` to `contentYPosition`

jojohoch committed:
Fix typo
